pa/cc/embedding.h

```diff
@@ -0,0 +1,272 @@
+    fill_ugraph(sigs1, sigs2);
+    p_graph_ = std::move(LabeledGraph<prop, prop>(pu_label, pv_label));
+    fill_pgraph();
+    fill_inv_label();
+  }
+
+  Graph& get_universe_graph() { return u_graph_; }
+  const Graph& get_universe_graph() const { return u_graph_; }
+  LabeledGraph<prop, prop>& get_predicate_graph() { return p_graph_; }
+  const LabeledGraph<prop, prop>& get_predicate_graph() const { return p_graph_; }
+  bool get_valid() const { return valid_; }
+
+  /* Assert that the constraint p_graph_.uLabel(pu) |-> p_graph_.vLabel(pv)
+     and performs "arc consistency" */
+  void choose_constraint(size_t pu, size_t pv, std::vector<Graph::VertexPair>& p_removed,
+			 std::vector<Graph::VertexPair>& u_removed){
+    /*    const std::vector<int>& u_vars = p_graph_.getULabel(pu).vars;
+    const std::vector<int>& v_vars = p_graph_.getVLabel(pv).vars;
+    for (size_t i = 0; valid_ && i < u_vars.size(); ++i){
+      valid_ = u_graph_.has_edge(u_vars[i], v_vars[i]);
+    }
+    if (!valid_){
+      return;
+    }
+    std::vector<Graph::VertexPair> tmp = u_graph_.remove_edges(u_vars, v_vars);
+    u_removed.insert(u_removed.end(), tmp.begin(), tmp.end());
+    std::vector<int> junk;
+    valid_ = u_graph_.unit_prop(u_removed, junk, junk); */
+
+    
+
+    std::vector<int> matches;
+    matches.resize(u_graph_.uSize(), -1);
+
+    std::vector<Graph::VertexPair> tmp;
+
+    std::vector<int> pu_units, pv_units;
+    pu_units.push_back(pu), pv_units.push_back(pv);
+    tmp = std::move(p_graph_.remove_edges(pu_units, pv_units));
+    p_removed.insert(p_removed.begin(), tmp.begin(), tmp.end());
+    if (!p_graph_.unit_prop(p_removed, pu_units, pv_units)) valid_ = false;
+
+    while (valid_ && pu_units.size() != 0){
+      std::vector<int> uu_units, uv_units;
+      for (size_t i = 0; valid_ && i < pu_units.size(); ++i){
+	const std::vector<int>& u_vars = p_graph_.getULabel(pu_units[i]).vars;
+	const std::vector<int>& v_vars = p_graph_.getVLabel(pv_units[i]).vars;
+	for (size_t k = 0; valid_ && k < u_vars.size(); ++k){
+	  if (matches[u_vars[k]] == -1){
+	    matches[u_vars[k]] = v_vars[k];
+	    uu_units.push_back(u_vars[k]);
+	    uv_units.push_back(v_vars[k]);
+	  } else if (matches[u_vars[k]] != v_vars[k]){ /* Universe Inconsistency */
+	    valid_ = false;
+	  }
+	}
+      }
+      if (!valid_) break;
+      tmp = std::move(u_graph_.remove_edges(uu_units, uv_units));
+      u_removed.insert(u_removed.begin(), tmp.begin(), tmp.end());
+      uu_units.clear(); uv_units.clear();
+      if (!u_graph_.unit_prop(u_removed, uu_units, uv_units)){
+	valid_ = false;
+	break;
+      }
+      pu_units.clear(); pv_units.clear();
+      //filter(uu_units, uv_units, p_removed, pu_units, pv_units);
+    }
+  }
+
+  /* Remove any edge inconsistent with U[i] |-> V[i] in the universe graph */
+  void filter(const std::vector<int>& U, const std::vector<int>& V,
```

Comment:
This filtering mechanism is ineffective, according to my limited experimentation -- it doesn't ever remove any edges in the examples I've tried (after uncommenting line 224).  I haven't tracked down why.